File: sdk/typescript/examples/workflow-automation.ts
/**
 * Workflow Automation Example
 *
 * Shows how to create automated workflows that chain debates
 * with conditional logic and external integrations.
 *
 * Usage:
 *   npx ts-node examples/workflow-automation.ts
 */

import { createClient } from '@aragora/sdk';

async function main() {
  const client = createClient({
    baseUrl: process.env.ARAGORA_API_URL || 'https://api.aragora.ai',
    apiKey: process.env.ARAGORA_API_KEY,
  });

  console.log('Creating workflow...');

  // Create a workflow with multiple steps
  const workflow = await client.workflows.create({
    name: 'Code Review Pipeline',
    description: 'Multi-stage code review with AI agents',
    triggers: [
      {
        type: 'webhook',
        config: { path: '/webhooks/code-review' },
      },
    ],
    steps: [
      {
        id: 'initial-review',
        type: 'debate',
        config: {
          task: 'Review this code for bugs and security issues: {{input.code}}',
          agents: ['claude', 'gpt-4'],
          rounds: 2,
        },
      },
      {
        id: 'check-severity',
        type: 'condition',
        config: {
          expression: 'steps.initial_review.consensus.confidence > 0.8',
          onTrue: 'deep-analysis',
          onFalse: 'quick-summary',
        },
      },
      {
        id: 'deep-analysis',
        type: 'debate',
        config: {
          task: 'Perform deep security analysis on: {{input.code}}',
          agents: ['claude', 'gemini'],
          rounds: 3,
          context: '{{steps.initial_review.consensus}}',
        },
      },
      {
        id: 'quick-summary',
        type: 'transform',
        config: {
          template: 'Quick review completed. Findings: {{steps.initial_review.consensus.final_answer}}',
        },
      },
      {
        id: 'notify',
        type: 'webhook',
        config: {
          url: '{{env.SLACK_WEBHOOK_URL}}',
          method: 'POST',
          body: {
            text: 'Code review complete: {{workflow.output}}',
          },
        },
      },
    ],
  });

  console.log(`Workflow created: ${workflow.id}`);

  // Execute the workflow with sample input
  const execution = await client.workflows.execute(workflow.id, {
    input: {
      code: `
        def authenticate(username, password):
            query = f"SELECT * FROM users WHERE username='{username}' AND password='{password}'"
            return db.execute(query)
      `,
    },
  });

  console.log(`Execution started: ${execution.execution_id}`);

  // Wait for completion
  let status = execution;
  while (status.status === 'running') {
    await new Promise((r) => setTimeout(r, 3000));
    status = await client.workflows.getExecution(execution.execution_id);
    console.log(`Status: ${status.status}, Step: ${status.current_step}`);
  }

  console.log('\n--- Workflow Results ---');
  console.log(JSON.stringify(status.output, null, 2));
}

main().catch(console.error);

File: sdk/python/examples/workflow_automation.py
"""
Workflow Automation Example

Shows how to create and execute workflow templates that chain debates
with conditional logic and external integrations.

Usage:
    python examples/workflow_automation.py

Environment:
    ARAGORA_API_KEY - Your API key
    ARAGORA_API_URL - API URL (default: https://api.aragora.ai)
"""

from __future__ import annotations

import os
import time

from aragora import AragoraClient


def main() -> None:
    client = AragoraClient(
        base_url=os.environ.get("ARAGORA_API_URL", "https://api.aragora.ai"),
        api_key=os.environ.get("ARAGORA_API_KEY"),
    )

    # List available workflow templates
    print("=== Available Templates ===\n")
    templates = client.workflows.list_templates(category="analysis")
    for tmpl in templates.get("templates", [])[:5]:
        print(f"  {tmpl['name']} ({tmpl['id']})")
        print(f"    {tmpl.get('description', 'No description')}")
        print()

    # Create a workflow from a pattern
    print("Creating workflow...\n")
    workflow = client.workflows.instantiate_pattern(
        pattern_id="code-review",
        name="Code Review Pipeline",
        description="Multi-stage code review with AI agents",
        category="security",
        agents=["claude", "gpt-4"],
    )

    workflow_id = workflow["workflow_id"]
    print(f"Workflow created: {workflow_id}")

    # Execute the workflow
    print("Executing workflow...\n")
    execution = client.workflows.execute(
        workflow_id=workflow_id,
        input_data="""
def authenticate(username, password):
    query = f"SELECT * FROM users WHERE username='{username}'"
    return db.execute(query)
""",
    )

    execution_id = execution["execution_id"]
    print(f"Execution started: {execution_id}")

    # Poll for completion
    status = execution
    while status.get("status") in ("running", "pending"):
        time.sleep(3)
        status = client.workflows.get_execution(execution_id)
        step = status.get("current_step", "unknown")
        print(f"  Status: {status['status']}, Step: {step}")

    # Print results
    print("\n--- Workflow Results ---")
    if status.get("status") == "completed":
        output = status.get("output", {})
        print(f"  Final output: {output}")
    else:
        print(f"  Workflow ended with status: {status.get('status')}")
        if status.get("error"):
            print(f"  Error: {status['error']}")


if __name__ == "__main__":
    main()
